spec/build_code_calculator_spec.rb

```diff
@@ -1,12 +1,23 @@
 require 'spec_helper'
 
 describe Fastlane::Wpmreleasetoolkit::Versioning::BuildCodeCalculator do
-  describe 'bumps the build code' do
-    it 'increments the build code by 1' do
-      build_code = Fastlane::Models::BuildCode.new(735)
-      calculator = described_class.new(build_code)
-      bumped_build_code = calculator.calculate_next_build_code
+  describe 'calculates the next build code' do
+    it 'increments the simple build code by 1' do
+      build_code = 735
+      bumped_build_code = described_class.new.next_simple_build_code(after: build_code)
       expect(bumped_build_code.to_s).to eq('736')
     end
+
+    it 'increments the derived build code by 1 with version numbers that are single digits' do
+      version = Fastlane::Models::AppVersion.new(1, 2, 3, 4)
+      bumped_build_code = described_class.new.next_derived_build_code(after: version)
+      expect(bumped_build_code.to_s).to eq('101020304')
+    end
+
+    it 'increments the derived build code by 1 with version numbers that are two digits' do
+      version = Fastlane::Models::AppVersion.new(12, 34, 56, 78)
+      bumped_build_code = described_class.new.next_derived_build_code(after: version)
+      expect(bumped_build_code.to_s).to eq('112345678')
+    end
```

Comment:
⚠️ @spencertransier   Oh wait, actually I just realized that it's not completely true for Tumblr… depending on the platform we're talking about.

 - For Android, currently the `versions.gradle` file contains all 4 individual components `major=31`, `minor=4`, `patch=0`, `number=108` for each `AppVersion` part, and then both the `versionName` and `versionCode` can be fully derived from those. Computing the next build code is a matter of incrementing 108 to 109, and the `versionCode` value will just be reconstructed via `1%<major>.2i%.1i<minor>%<patch>.1i%<number>.3i`.
 - For iOS, the `VERSION_SHORT` looks like just `31.4` while the `VERSION_LONG` looks like `314008` (as in `%<major>.2i%<minor>.1i%<patch>.1i%<build_number>.2i`)

So that means that:
1. The way the `BuildCode` is derived differs from Android vs iOS. We'll likely need two subclasses of `BuildCodeCalculator`, one for each case. So maybe my suggestion from the other comment about making those subclasses will have to be implemented after all
2. You were right that, for the iOS case at least, the derived build code can't be fully determined by just the `AppVersion` and instead still have this concept of being "incremented" from previous `BuildCode` version after all too

And also, another thing that I noticed while looking at WordPress iOS, is that while some other apps use just an incrementing number for the `VERSION_LONG`, WPiOS just uses the 4-components-with-dots format right away (e.g. `23.2.2.0`). In fact the Build Code for iOS doesn't seem to have to be an integer.

---

That means we might have to rethink that concept of `BuildCode` after all.

 - For Android it **_has_** to be an `Integer`
    - Some projects are just incrementing it by 1 continuously, even after a `<major>.<minor>` version bump happened
    - Other projects, like Tumblr, are incrementing it by 1 when doing just a new build on the same major.minor… but recomputing a new base / starting-point `Integer` value to start from every time we bump `<major>.<minor>.<patch>` to a new version (so `versionCode` for `31.2.1` will start at `1310201100` and increment from there, until we do a code freeze and bump the release version to `31.3.0`, where the next `versionCode` will start at `1310300100`, leaving enough room for hotfixes and patches that might still need to happen on `31.2.*`
 - For iOS, it can be an `Integer`… but also can just as well be a dot-separated `String`
    - WPiOS and WCiOS seem to use `major.minor.patch.number` directly for example
    - Tumblr currently uses an Integer, but that is derived from the same, just without the dots (e.g. `315001` instead of `31.5.0.01`). I wouldn't mind us switching Tumblr-iOS to use dots in their `VERSION_LONG` though tbh, i.e. `31.5.0.1` instead of `315001`, now that I realized it doesn't have to be an `Integer` — that way we won't need any special casing for Tumblr-iOS, and besides that will make it way more readable
    - I'm not sure what PCiOS and DOiOS use?

---

Overall that means that:
 - iOS projects will only need `VersionCalculator` and manipulate `AppVersion` models. From those `AppVersion`, the `VersionFormatter` can derive both the `VERSION_SHORT` (`<major>.<minor>.<patch if ≠ 0>`) and the `VERSION_LONG` (`<major>.<minor>.<patch>.<build_number>`), and they don't really need the concept of `BuildCode`.
 - Android projects will need both a `VersionCalculator` and a `BuildCodeCalculator`. The latter can have a simple method like `next_code_after(version:, code:)`, implemented slightly differently for its 2 subclasses:
    - For the calculator used by WP, WC, …, that method will just return `code+1` (and ignore the value of `version`)
    - For the one used by Tumblr, that method will check that `code` format starts with the proper prefix corresponding to the `version`'s `major`, `minor` and `patch`. If it does, it will just return `code+1`… but if not, that means that we've bumped the major/minor/patch and we need to use a new base, so it will return `format('1%<major>02i%<minor>02i%<patch>02i100')` aka the first `BuildCode` to use for that new version (Tumblr starts their build numbers of each new version at 100 for some legacy reason)

This is just my raw braindump 🧠  of my live reflexions as I'm trying to think of all the cases (iOS vs Android, Tumblr vs WP/WC vs DO vs PC), so we might also need deeper reflexions (and think how the call sites will work out for each of those platforms & apps) and taking a step back before starting changing the implementation again 🙃